File: game_server/Common/src/NPCommon/Game/CS/CSWeightRandomList.java
package NPCommon.Game.CS;

import NPCommon.Log.CommLog;
import NPCommon.RefData._IParseFromStringable;

import java.util.ArrayList;
import java.util.List;

/**
 * 权重随机对象
 */
public class CSWeightRandomList implements _IParseFromStringable
{
    private List<CSWeightItem> _m_items;
    private int _m_totalWeight;

    public CSWeightRandomList()
    {
        this._m_items = new ArrayList<>();
        this._m_totalWeight = 0;
    }

    /**
     * 从字符串解析权重配置
     * 格式: "数据1id:权重1;数据2id:权重2"
     */
    public boolean parseFromString(String _config)
    {
        _m_items.clear();
        _m_totalWeight = 0;

        if (_config == null || _config.trim().isEmpty())
        {
            return true;
        }

        String[] _pairs = _config.split(";");
        for (String _pair : _pairs)
        {
            if (_pair.trim().isEmpty())
            {
                continue;
            }

            String[] _parts = _pair.split(":");
            if (_parts.length != 2)
            {
                return false;
            }

            try
            {
                long _dataId = Long.parseLong(_parts[0].trim());
                int _weight = Integer.parseInt(_parts[1].trim());

                addItem(_dataId, _weight);
            } catch (NumberFormatException e)
            {
                return false;
            }
        }

        return true;
    }

    /**
     * 添加权重项
     */
    public void addItem(long _dataId, int _weight)
    {
        if (_weight <= 0)
        {
            CommLog.warn("WeightRandomList addItem Invalid weight:{}", _weight, new Exception());
            return;
        }

        _m_items.add(new CSWeightItem(_dataId, _weight));
        _m_totalWeight += _weight;
    }

    /**
     * 使用权重随机选择一个数据ID
     */
    public Long randomSelect(CSSyncRandom _random)
    {
        if (_m_items.isEmpty())
            return null;

        int _randomValue = _random.nextInt(_m_totalWeight);
        int _currentWeight = 0;

        for (CSWeightItem _item : _m_items)
        {
            _currentWeight += _item.getWeight();
            if (_randomValue < _currentWeight)
            {
                return _item.getDataId();
            }
        }

        // 理论上不应该到达这里
        return _m_items.get(_m_items.size() - 1).getDataId();
    }

    /**
     * 获取所有权重项
     */
    public List<CSWeightItem> getItems()
    {
        return new ArrayList<>(_m_items);
    }

    /**
     * 获取总权重
     */
    public int getTotalWeight()
    {
        return _m_totalWeight;
    }

    /**
     * 检查是否为空
     */
    public boolean isEmpty()
    {
        return _m_items.isEmpty();
    }

    /**
     * 清空所有权重项
     */
    public void clear()
    {
        _m_items.clear();
        _m_totalWeight = 0;
    }
}

File: game_client/Assets/Scripts/ResScripts/NPSO/GSO/CommonInfo/CSWeightRandomList.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace GOE
{
    /// <summary>
    /// 权重随机对象
    /// </summary>
    [Serializable]
    public class CSWeightRandomList : _IParseFromStringable
    {
        [SerializeField]
        private List<CSWeightItem> _m_lItems;
        private int _m_lTotalWeight;

        public CSWeightRandomList()
        {
            _m_lItems = new List<CSWeightItem>();
            _m_lTotalWeight = 0;
        }

        /// <summary>
        /// 从字符串解析权重配置
        /// 格式: "数据1id:权重1;数据2id:权重2"
        /// </summary>
        public void ParseFromString(string config)
        {
            _m_lItems.Clear();
            _m_lTotalWeight = 0;

            if (string.IsNullOrWhiteSpace(config))
            {
                return;
            }

            string[] pairs = config.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string pair in pairs)
            {
                if (string.IsNullOrWhiteSpace(pair))
                {
                    continue;
                }

                string[] parts = pair.Split(':');
                if (parts.Length != 2)
                {
                    return;
                }

                try
                {
                    long dataId = long.Parse(parts[0].Trim());
                    int weight = int.Parse(parts[1].Trim());

                    AddItem(dataId, weight);
                }
                catch (Exception e)
                {
                    Debug.LogError($"CSWeightRandomList ParseFromString error:{e.ToString()}");
                    return;
                }
            }

            return;
        }

        /// <summary>
        /// 添加权重项
        /// </summary>
        public void AddItem(long dataId, int weight)
        {
            if (weight <= 0)
            {
                Debug.LogWarning($"WeightRandomList addItem Invalid weight: {weight}");
                return;
            }

            _m_lItems.Add(new CSWeightItem(dataId, weight));
            _m_lTotalWeight += weight;
        }

        /// <summary>
        /// 使用权重随机选择一个数据ID
        /// </summary>
        public long RandomSelect(CSSyncRandom random)
        {
            if (_m_lItems.Count == 0)
                return 0;

            int randomValue = random.NextInt(TotalWeight);
            int currentWeight = 0;

            foreach (CSWeightItem item in _m_lItems)
            {
                currentWeight += item.weight;
                if (randomValue < currentWeight)
                {
                    return item.dataId;
                }
            }

            // 理论上不应该到达这里
            return _m_lItems[_m_lItems.Count - 1].dataId;
        }

        /// <summary>
        /// 获取所有权重项
        /// </summary>
        public List<CSWeightItem> Items
        {
            get { return new List<CSWeightItem>(_m_lItems); }
        }

        /// <summary>
        /// 获取总权重
        /// </summary>
        public int TotalWeight
        {
            get
            {
                if (_m_lTotalWeight <= 0)
                {
                    for (int i = 0; i < _m_lItems.Count; i++)
                    {
                        _m_lTotalWeight += _m_lItems[i].weight;
                    }
                }
                return _m_lTotalWeight;
            }
        }

        /// <summary>
        /// 检查是否为空
        /// </summary>
        public bool IsEmpty
        {
            get { return _m_lItems.Count == 0; }
        }

        /// <summary>
        /// 清空所有权重项
        /// </summary>
        public void Clear()
        {
            _m_lItems.Clear();
            _m_lTotalWeight = 0;
        }
    }
} 